src/MessageTrait.php

```diff
@@ -327,13 +330,18 @@ private function getStream($stream, string $modeIfNotInstance): StreamInterface
      */
     private function setHeaders(array $originalHeaders): void
     {
+        if ($this->arrayIsList($originalHeaders)) {
```

Comment:
Do we need this check?

Is `0` disallowed in the RFC mentioned in the comments or why are numeric-only headers not allowed?
and if `0` is disallowed, why is `-1`, `0`, `1` allowed (which is not a list but still containers only numeric keys)?

Reply:
This is based on a polyfill for `array_is_list()` that mimics the behavior of that function... which only returns true if the keys are integers, are in sequence, and the sequence starts with 0. As such, keys starting at -1 will cause that function to return `false`.

All that said... I've noted on the Slack my reservations about the approach in this patch, and have asked for some guidance on an alternate approach.